File: problems/Java/84. Largest Rectangle in Histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        Deque<int[]> stk = new ArrayDeque<>();
        int area = 0;
        
        for (int i = 0; i < heights.length; i++) {
            int start = i;
            while (!stk.isEmpty() && stk.peek()[1] > heights[i]) {
                int[] bar = stk.pop();
                int idx = bar[0];
                int hgt = bar[1];
                area = Math.max(area, hgt * (i - idx));
                start = idx;
            }
            stk.push(new int[]{start, heights[i]});
        }
        
        while (!stk.isEmpty()) {
            int[] bar = stk.pop();
            int idx = bar[0];
            int hgt = bar[1];
            area = Math.max(area, hgt * (heights.length - idx));
        }
        
        return area;
    }
}

File: problems/Python/84. Largest Rectangle in Histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stk = []
        area = 0
        
        for i, h in enumerate(heights + [0]):
            start = i
            while stk and stk[-1][1] > h:
                idx, hgt = stk.pop()
                area = max(area, hgt * (i - idx))
                start = idx
            stk.append((start, h))
            
        return area